src/h5web/visualizations/nexus/utils.ts

```diff
@@ -46,30 +46,48 @@ export function getLinkedEntity(
 export function getNxAxes(
   group: HDF5Group,
   metadata: HDF5Metadata
-): { labels: Array<string | undefined>; ids: Record<string, HDF5Id> } {
+): {
+  names: Array<string | undefined>;
+  labels: Record<string, string>;
+  ids: Record<string, HDF5Id>;
+} {
   const axesList = getAttributeValue(group, 'axes');
 
   if (!axesList) {
-    return { labels: [], ids: {} };
+    return { names: [], labels: {}, ids: {} };
   }
 
   const axes = typeof axesList === 'string' ? [axesList] : axesList;
   assertArray<string>(axes);
 
-  const axesLabels = axes.map((a) => (a === '.' ? undefined : a));
-  const axesIds = axesLabels.reduce<Record<string, HDF5Id>>((acc, axis) => {
-    if (!axis) {
-      return acc;
-    }
+  const axesNames = axes.map((a) => (a === '.' ? undefined : a));
+  const axesDatasets = axesNames.reduce<Array<[string, HDF5Dataset]>>(
+    (acc, axis) => {
+      if (!axis) {
+        return acc;
+      }
 
-    const dataset = getLinkedEntity(group, metadata, axis);
-    if (dataset && isDataset(dataset)) {
-      acc[axis] = dataset.id;
-    }
-    return acc;
-  }, {});
+      const dataset = getLinkedEntity(group, metadata, axis);
+      if (dataset && isDataset(dataset)) {
+        acc.push([axis, dataset]);
+      }
+      return acc;
+    },
+    []
+  );
 
-  return { labels: axesLabels, ids: axesIds };
+  return {
+    names: axesNames,
+    labels: Object.fromEntries(
+      axesDatasets.map(([axis, dataset]) => [
+        axis,
+        getDatasetLabel(dataset, axis),
+      ])
+    ),
+    ids: Object.fromEntries(
+      axesDatasets.map(([axis, dataset]) => [axis, dataset.id])
+    ),
+  };
```

Comment:
`axesNames` are the dataset names as stored in `@axes` (previously called `axesLabels`)

`axesLabels` are now an object with dataset names as keys and labels (which can now be different) as values.

Reply:
I feel like we're starting to have a lot of variables objects/arrays that are coupled to one another. I think we need a better data structure. Maybe this function would return the `axesParams` object I mention above: 

```ts
interface AxisParams {
  name: string;
  label: string;
  values: HDF5Value | undefined;
}

type AxesParams = Record<id, AxisParams>;
```

Not sure exactly, I'd have to look at it more, but do you get my point?

Reply:
I could also imagine the `useDatasetValues` hook being refactored to have a signature like:

```ts
interface VisParams {
  title?: string;
  values: HDF5Value | undefined;
  axesParams: AxesParams;
}

function useVisParams(signalId: string, axesParams: AxesParams): VisParams | undefined {...}
```

The mapped vis would then pick the correct axis params object for the abscissa/ordinate based on `mapperState` as they do now.

Anyway, just some thoughts. If you prefer to do this later, let me know, but I think this PR is a good opportunity.